Update OpenTelemetry packages and use built-in metric methods

martincostello opened the pull request:
- Update all the OpenTelemetry packages to their latest versions.
- Use built-in methods to add .NET metrics.
- Add suppression for new AoT warning on `AddSqlClientInstrumentation()`.

Resolves #1945.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/1948)

martincostello committed:
Update OpenTelemetry packages

- Update all the OpenTelemetry packages to their latest versions.
- Use built-in methods to add .NET metrics.
- Add suppression for new AoT warning on `AddSqlClientInstrumentation()`.

martincostello commented on playground/orleans/OrleansServiceDefaults/Extensions.cs:
```diff
@@ -42,8 +42,10 @@ public static IHostApplicationBuilder ConfigureOpenTelemetry(this IHostApplicati
+                metrics.AddAspNetCoreInstrumentation()
+                       .AddHttpClientInstrumentation()
+                       .AddRuntimeInstrumentation()
+                       .AddMeter("Microsoft.Orleans");
```
I figured it was nicer to hoist this out for just the one meter, but I can put it back as it was if that's preferred (unless this has its own extension method now too).

martincostello commented on src/Components/Aspire.Microsoft.Data.SqlClient/AspireSqlServerSqlClientExtensions.cs:
```diff
@@ -89,7 +89,9 @@ string GetConnectionString()
         {
             builder.Services.AddOpenTelemetry().WithTracing(tracerProviderBuilder =>
             {
+#pragma warning disable IL2026 // Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code
                 tracerProviderBuilder.AddSqlClientInstrumentation();
+#pragma warning restore IL2026 // Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code
```
I tried guarding this with `RuntimeFeature.IsDynamicCodeSupported`, but it didn't seem to help.

samsp-msft replied:
What is your goal here? Has SqlClient been made trim or nativeAOT safe?
The right way to solve trimming warnings is by [annotating](https://learn.microsoft.com/dotnet/core/deploying/trimming/prepare-libraries-for-trimming?pivots=dotnet-8-0#resolve-trim-warnings) the methods involved, not directly disabling warnings.

martincostello replied:
I presumed that this code is intending to be trim compatible and _appeared_ to be previously but only because there was nothing indicating it wasn't. The extension method from OTel is now explicitly annotated as not trim friendly.

Possibilities I can see are:

1. This propagates outwards virally from here and makes everything using this code not AoT compatible.
2. The method is only called when not running under AoT (which still needs a suppression as I tried that and still got an error).
3. Suppress it on the basis it doesn't work but that's a problem for later.
4. Not upgrade the package that introduces the warning, but that's likely a false negative as it probably didn't work before, it just didn't have the annotation to say that.

1 is the most disruptive but most technically correct, but I don't know of the context here to know if that would be too breaking.

Ultimately the motivation was "stop the build failing", I'm happy to change it to whatever is the preferred resolution.

eerhardt replied:
This component is not trim-compatible because its dependencies aren't trim compatible.

We shouldn't be making this change. But instead, we should add:

```xml
<IsAotCompatible>false</IsAotCompatible>
```

to the .csproj.

samsp-msft replied:
Should that be part of the this PR or a different one?

eerhardt replied:
If we are updating the version of the OTel package in this PR, some change needs to be here because the new version is causing our build to fail.

samsp-msft commented:
Now that the OTel library has[ conditional logic ](https://github.com/open-telemetry/opentelemetry-dotnet/blob/0b9cf42e9af668fe6d3af86254f9758c7c821f5e/src/OpenTelemetry.Instrumentation.AspNetCore/AspNetCoreInstrumentationMeterProviderBuilderExtensions.cs#L42-L51)to use the meters in .NET 8, overall this seems like a worthy change.

@eerhardt - what's the right fix for the trimming issue?
@joperezr - do we have a process for evaluating updating dependency package versions as part of the preview release process?

martincostello committed:
Resolve review feedback

Mark as not AoT compatible due to dependency on OpenTelemetry.Instrumentation.SqlClient, which itself is not AoT compatible.

davidfowl commented:
/azp run

davidfowl commented:
Thanks @martincostello ! Solid work as usual
